decred/decred/dcr/account.py

```diff
@@ -831,19 +1088,169 @@ def calcBalance(self, tipHeight=None):
+        """
+        stakebases = sum(ticket.tinfo.stakebase for ticket in self.ticketDB.values())
+        poolFees = sum(ticket.tinfo.poolFee for ticket in self.ticketDB.values())
+        txFees = sum(
+            ticket.tinfo.purchaseTxFee + ticket.tinfo.spendTxFee
+            for ticket in self.ticketDB.values()
+        )
+        return stakebases, poolFees, txFees
+
     def updateStakeStats(self):
         """
         Updates the stake stats object.
         """
-        ticketCount = 0
-        ticketVal = 0
-        for utxo in self.utxos.values():
-            if utxo.isTicket():
-                ticketCount += 1
-                ticketVal += utxo.satoshis
-                self.tickets.append(utxo.txid)
+        uT = self.unspentTickets()
+        ticketCount = len(uT)
+        ticketVal = sum(t.satoshis for t in uT.values())
         self.stakeStats = TicketStats(ticketCount, ticketVal)
 
+    def spentTicketInfo(self, txid):
+        """
+        It's possible that when retrieving the tinfo from dcrdata that it is not
+        ready. Log a message and return None if so. Return the TicketInfo if found.
+
+        Args:
+            txid (str): The ticket to find info for.
+
+        Returns:
+            TicketInfo: The ticket info if found or None.
+        """
+        tinfo = self.blockchain.ticketInfo(txid)
+        if tinfo.vote or tinfo.revocation:
+            return tinfo
+        log.debug("ticket's tinfo not found on the blockchain {}".format(txid))
+
+    def updateTicket(self, tx):
+        """
+        Update a ticket that is already in the ticketDB. Sends a
+            ui.SPENT_TICKETS_SIGNAL on completion.
+
+        Args:
+            tx (msgtx.MsgTx): the vote or transaction that spends a ticket in
+                the ticketDB.
+        """
+        isVote = len(tx.txIn) > 1
+        idx = 1 if isVote else 0
+        ticketTxid = tx.txIn[idx].previousOutPoint.txid()
+        if ticketTxid not in self.ticketDB:
+            raise Exception(
+                "spending tx came for ticket that was not found in the database: {}".format(
+                    ticketTxid
+                )
+            )
+        # sometimes this data cannot be fetched yet by dcrdata
+        tinfo = self.spentTicketInfo(ticketTxid)
+        if not tinfo:
+            if isVote:
+                tinfo = TicketInfo.fromVote(self.blockchain, tx)
+            else:
+                tinfo = TicketInfo.fromRevocation(self.blockchain, tx)
+
+        ticket = self.ticketDB[ticketTxid]
+        ticket.tinfo = tinfo
+        self.ticketDB[ticketTxid] = ticket
+
+        self.signals.spentTickets()
+
+    def updateSpentTickets(self):
+        """
+        Update voted and revoked tickets. ui.SPENT_TICKETS_SIGNAL is called if
+        there was an update.
+        """
+        # live tickets
+        mempoolTickets = self.mempoolTickets()
+        tickets = self.liveTickets()
+        tickets.update(mempoolTickets)
```

Comment:
Don't you want immature tickets here too?

Reply:
I think the method is misleading. It is used to find live tickets as in current on the blockchain, unspent with varying statuses.  Since It and mempool tickets should only be used to update the ticket database, and other methods should use the ticket database to find information about a ticket, I localized these functions to updateSpentTickets. Hope that's ok.